src/harness/reference_models/pre_iap_filtering/fss_purge.py

```diff
@@ -0,0 +1,301 @@
+      purged and corresponding cbsd data object respectively.
+  """
+  # List of grants to be purged for the fss_entity.
+  final_purge_list = []
+
+  # List of dictionaries containing the highest grant of the CBSD and a reference to the CBSD.
+  grants_cbsds_info_for_oobe_calculation = getHighestGrantOfCbsds(cbsds)
+
+  # Calculate the threshold value in dBm for the OOBE interference
+  oobe_threshold_value = -129 + 10 * (np.log10(REF_BW/interference.ONE_MHZ) - FSS_OOBE_MARGIN)
+
+  while True:
+    # Calculate the OOBE interference value for each grant.
+    calculateOobeInterference(grants_cbsds_info_for_oobe_calculation, fss_entity)
+
+    # Sort the grants based on the calculated interference value.
+    sorted_grants_cbsds_info = sorted(grants_cbsds_info_for_oobe_calculation, key=lambda grant: grant['oobe_interference'])
+
+    # Find the largest index such that aggregation of the interferences do not cross
+    # the OOBE threshold.
+    oobe_interferences = [interference.dbToLinear(grant['oobe_interference']) for grant in sorted_grants_cbsds_info]
+    cumulated_interference = np.cumsum(oobe_interferences)
+    index = np.searchsorted(cumulated_interference, interference.dbToLinear(oobe_threshold_value))
+    grants_to_purge = sorted_grants_cbsds_info[index:]
+    grants_cbsds_info_for_oobe_calculation = sorted_grants_cbsds_info[:index]
+    for grants_cbsds in grants_to_purge:
+      final_purge_list.append(grants_cbsds_namedtuple(grants_cbsds['grantData'], grants_cbsds['cbsd']))
+
+    # Add the identified grants to be purged from this iteration to the final list of grants to be purged
+    if grants_to_purge:
+      for purged_grant in grants_to_purge:
+        remaining_grants = []
+        # Identify other grants of the CBSD that have the same emission mask as the purged
+        # grant and add them to the final purge list. If the CBSD has any other grant with
+        # lower emission mask then add it to the remaining_grants list.
+        for grant in purged_grant['cbsd']['grants']:
+          # Check that we are not comapring the same grant as the pruged grant
+          if grant != purged_grant['grantData']:
+            if (getMcbsdValue(purged_grant['grantData']) == getMcbsdValue(grant)):
+              final_purge_list.append(grants_cbsds_namedtuple(grant, purged_grant['cbsd']))
+            else:
+              remaining_grants.append(grant)
+
+        # Identify the grant with the highest highFrequncy value among the remaining grants for the
+        # CBSD and add it to the grants_cbsds_info_for_oobe_calculation list for next iteration.
+        if len(remaining_grants) > 0:
+          max_grant = max(remaining_grants,\
+                           key=lambda x: x['operationParam']['operationFrequencyRange']['highFrequency'])
+          grants_cbsds_info_for_oobe_calculation.append({'grantData': max_grant, 'cbsd': purged_grant['cbsd']})
```

Comment:
Shouldn't we remove the purged grants before appending the new grants into grants_cbsds_info_for_oobe_calculation

Reply:
I think the current implementation is correct (barring previously mentioned bugs), grants_cbsds_info_for_oobe_calculation is all the CBSD-grant tuples that are not in the purgelist for this iteration. The next iteration must consider both those grants/CBSDs and any grants from CBSDs that also have a grant at a lower emission mask, which this is supposed to be adding.

Reply:
I see that now. Missed the line where the assignment is being done